polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+        Operator::Mul => unimplemented!("multiplication is not supported for data filtering."),
+        Operator::Div => unimplemented!("division is not supported for data filtering."),
+        Operator::Mod => unimplemented!("`mod` is not supported for data filtering."),
+        Operator::Rem => unimplemented!("`rem` is not supported for data filtering."),
+        Operator::Add => unimplemented!("addition is not supported for data filtering."),
+        Operator::Sub => unimplemented!("subtraction is not supported for data filtering."),
+        Operator::Geq => unimplemented!("`>=` is not supported for data filtering."),
+        Operator::Leq => unimplemented!("`<=` is not supported for data filtering."),
+        Operator::Neq => unimplemented!("`!=` is not supported for data filtering."),
+        Operator::Gt => unimplemented!("`>` is not supported for data filtering."),
+        Operator::Lt => unimplemented!("`<` is not supported for data filtering."),
+        // @TODO(steve): Expand or expressions to multiple bindings in the simplifier.
+        Operator::Or => unimplemented!("`or` is not supported for data filtering."),
+        Operator::ForAll => unimplemented!("`forall` is not supported for data filtering."),
+    }
+    None
+}
+
+#[derive(Debug)]
+struct Vars {
+    variables: HashMap<String, HashSet<Symbol>>,
+    field_relationships: HashSet<(String, String, String)>,
+    in_relationships: HashSet<(String, String)>,
+    eq_values: HashMap<String, Term>,
+    contained_values: HashMap<String, HashSet<Term>>,
+    types: HashMap<String, String>,
+    this_id: String,
+}
+
+/// Collapses the var info that we obtained from walking the expressions.
+/// Track equivalence classes of variables and assign each one an id.
+fn collapse_vars(var_info: VarInfo) -> Vars {
+    // Merge variable cycles.
+    let mut joined_cycles: Vec<HashSet<Symbol>> = vec![];
+    'cycles: for (l, r) in var_info.cycles {
+        // See if we can add to an existing cycle.
+        for joined_cycle in &mut joined_cycles {
+            if joined_cycle.contains(&l) || joined_cycle.contains(&r) {
+                joined_cycle.insert(l);
+                joined_cycle.insert(r);
+                continue 'cycles;
+            }
+        }
+        // Create new cycle if we couldn't
+        let mut new_cycle = HashSet::new();
+        new_cycle.insert(l);
+        new_cycle.insert(r);
+        joined_cycles.push(new_cycle);
+    }
+
+    let mut next_id = 0;
+    let mut get_id = move || {
+        let id = next_id;
+        next_id += 1;
+        format!("{}", id)
+    };
+
+    // Give each cycle an id
+    let mut variables: HashMap<String, HashSet<Symbol>> = HashMap::new();
+    for joined_cycle in joined_cycles {
+        let id = get_id();
+        variables.insert(id, joined_cycle);
+    }
+
+    // Substitute in relationships
+    let mut parent_ids = vec![];
+    'parent_relationships: for (parent, _, _) in &var_info.field_relationships {
```

Comment:
Note to self: `field_relationships` is (var, field, other_var) (var.field = other_var)